### PR TITLE
Feature: Group functionality

### PR DESCRIPTION
**Pull request by SteffenMeinecke**

### CODE REVIEW

**Comment by SteffenMeinecke** [comment]
@rbolgaryn, the test `test_sqlite_workaround()` fails. Obviously, `from_sqlite()` cannot handle the new net dataframe `group` for some reason. The fail occurs at file_io.py line 391 `table = pd.read_sql_query("SELECT * FROM %s" % t, con, index_col="index")`. Since I am not an expert in SQL maybe you or @ascheidl have an idea?

**Comment by SteffenMeinecke** [comment]
regarding Group class i/o: @rbolgaryn should we move some functionality to io_utils.py or is it ok leave it in groups.py?

PS: @rbolgaryn maybe you have an idea how to improve `Group_to_json()` to not need this code:
```
{"_module": "pandapower.group", '
 '"_class": "Group", "_object": %s}' ...
```

**Comment by ascheidl** [comment]
> @rbolgaryn, the test `test_sqlite_workaround()` fails. Obviously, `from_sqlite()` cannot handle the new dataframe `net.group` for some reason. The fail occurs at file_io.py line 391 `table = pd.read_sql_query("SELECT * FROM %s" % t, con, index_col="index")`. Since I am not an expert in SQL maybe you or @ascheidl have an idea?

group is a sql literal ... it should work if you escape the table name:

 `table = pd.read_sql_query("SELECT * FROM '%s'" % t, con, index_col="index")`

**Review by rbolgaryn — CHANGES_REQUESTED** [review]
I think we should change the variable names to more human-readable (e.g. elms -> elements, elm_col -> element_column). This would also make the code more consistent with the code in the controller module.

set_element_column()
elements_dict
get_index
count_elements()
update_elements_dict()

**Comment by rbolgaryn** [comment]
I think the IO part of the code should be moved to the file that handles IO for other objects. 
I think the funkctions group_to/from_json(string) should not be necessary, the pp.to_json / pp.to_json should be able to handle it.

**Comment by rbolgaryn** [comment]
We can also define the functions set_in_service, set_out_of_service as properties instead, so that one can use 

```python
group1.in_service = False
group1.in_service = True
```

What do you think?

**Comment by SteffenMeinecke** [comment]
> We can also define the functions set_in_service, set_out_of_service as properties instead, so that one can use
> 
> ```python
> group1.in_service = False
> group1.in_service = True
> ```

IMHO, this is not possible because `Group.set_in_service(self, net)` needs `net` as input.

**Comment by SteffenMeinecke** [comment]
> I think the IO part of the code should be moved to the file that handles IO for other objects. I think the funkctions group_to/from_json(string) should not be necessary, the pp.to_json / pp.to_json should be able to handle it.

To register `Group` for `to_serializable` and `from_serializable` `Group` must be available. However, `Group` inherits from `JSONSerializableClass`. This would lead to a ring import if `JSONSerializableClass` is not moved to a file other than `io_utils.py`. Furthermore, I tried the idea of replacing the functionality of `Group_to_json()` and `Group_from_json()` but failed. @rbolgaryn could you do trial if you think it is really rational to replace the IO part from `groups.py`?

File: pandapower/groups.py [hunk 1]
```diff
@@ -0,0 +1,464 @@
...
+
+    def res_p_mw(self, net):
+        return self._sum_powers(net, "p", "mw")
+
+    def res_q_mvar(self, net):
+        return self._sum_powers(net, "q", "mvar")
+
+# --- i/o functionality
+
+@to_serializable.register(Group)
```
Comment (rbolgaryn) [inline pandapower/groups.py:436]: @SimonRubenDrauz please take a look at the I/O, is this okay from your point of view?

**Review by rbolgaryn — CHANGES_REQUESTED** [review]

File: pandapower/test/api/test_group.py [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+def typed_set(iterable, dtype):
+    if pd.api.types.is_integer_dtype(dtype):
+        return {int(it) for it in iterable}
+    elif pd.api.types.is_numeric_dtype(dtype):
+        return {float(it) for it in iterable}
+    else:
+        return {str(it) for it in iterable}
+
+
+def test_groups():
```
[the remark this answers is not in the document]
Comment (SteffenMeinecke) [reply pandapower/test/api/test_group.py]: 50c46140e006e726320640c26f53af9ee374d247 should solve this

**Comment by SteffenMeinecke** [comment]
from my point of view this PR is ready to merge.
Variant (2a) also makes sense from my point of view and can be implemented with less changes than (3). If we are of the opinion that (2a) fits much better into the pandapower user experience than the current implementation (1) and someone can be found who has the time to implement it, (2a) certainly makes sense as well.

**Comment by rbolgaryn** [comment]
Thank you for the overview @SteffenMeinecke !

I think the option 2.2 can also be implemented without the dictionary but by using the "long form": name, element, element_index.

**Comment by SteffenMeinecke** [comment]
> Thank you for the overview @SteffenMeinecke !
> 
> I think the option 2.2 can also be implemented without the dictionary but by using the "long form": name, element, element_index.

Since one line of the dataframe `net.group` corresponds to one group, I am not sure how you think to collect multiple elements of multiple element types by a "long form".

**Comment by ascheidl** [comment]
a little late, but I like the third option ... it seems to fit the tabular form of pandapower

**Comment by rbolgaryn** [comment]
> > Thank you for the overview @SteffenMeinecke !
> > I think the option 2.2 can also be implemented without the dictionary but by using the "long form": name, element, element_index.
> 
> Since one line of the dataframe `net.group` corresponds to one group, I am not sure how you think to collect multiple elements of multiple element types by a "long form".

There would be multiple lines describing 1 group in that case. To work with the group, one would query the group data, e.g. like this:

```python
def set_group_in_service(net, group, in_service):
    group_tab = net.group.loc[net.group.name==group]
    group_elements = group_tab.element.unique()
    for element in group_elements:
        if "in_service" not in net[element].columns:
            continue
        idx = group_tab.loc[group_tab.element == element].index
        net[element].loc[idx, "in_service"] = in_service
```

**Comment by SteffenMeinecke** [comment]
@rbolgaryn, ok, now I see your idea. But I'm not sure how the `.index` in the second last code line should work. I guess it should be rather `idx = group_tab.loc[group_tab.element == element].element_index` and `net.group.element_index` is of dtype list?

Furthermore, it should be noted that using the functionality of variable `element_column`, which I need since dataframe indices are changed in several subsequent functions, the functionality looks a bit more complex (as in current class implementation, the last if-else-condition of the following would be picked up by get_idx() and the whole function would be a special case of set_value()):
```
def set_out_of_service(net, group_name):
    group_tab = net.group.loc[net.group.name == group_name]
    dupl_elm_types = list(group_tab.element_type.loc[
        group_tab.element_type.duplicated()].unique())
    if len(dupl_elm_types):
        raise ValueError(f"Group {group_name} has duplicated entries to these element types: {dupl_elm_types}.")
    for idx in group_tab.index:
        et = group_tab.element_type.at[idx]
        if "in_service" not in net[et].columns:
            continue
        ec = group_tab.element_column.at[idx]
        eidx = group_tab.element_index.at[idx]
        if ec is None:
            net[et].loc[eidx, "in_service"] = False
        else:
            net[et].loc[net[et][ec].isin(eidx), "in_service"] = False
```
In that case, I would propose to put the group functions into pp.group so that the function names are easy-to-understand:
* pp.group.set_in_service(net, group_name)
* pp.group.set_out_of_service(net, group_name)
* pp.group.set_value(net, group_name, value, column, replace=True, append_column=True)
* pp.group.get_index(net, group_name, element_type)
* pp.group.get_elements_dict(net, group_name)
* pp.group.set_element_column(net, group_name, element_type=None)
* pp.group.drop(net, group_name, drop_elements=False)
* pp.group.drop_from_group(net, group_name, element_type=None, element_index=None, elements_dict=None)
* pp.group.append_to_group(net, group_name, element_type=None, element_index=None, elements_dict=None)
* ...
while creating a group is available in pp in file `create.py`: `pp.create_group(net, group_name, elements_dict, element_column=None, overwrite=False)`.

**Comment by ascheidl** [comment]
why not just pp.set_group_in_service ...

**Comment by rbolgaryn** [comment]
I think the element_index can refer to a single row, similarly to how it is implemented in net.poly_cost

**Comment by SteffenMeinecke** [comment]
ok, then I see this options:

* (1) **class implementation**, as already available
* (2) **group assignment in net.group dataframe**
    * (a) `net.group.dtypes`: `name: str`, `elements_dict: dict[str, pd.Index]`, `element_column: OPTIONAL[str]` (most similar to current implementation
    * (b) `net.group.dtypes`: `name: str`, `element_type: str`, `element_index: pd.Index`, `element_column: OPTIONAL[str]` (as I understood 17h ago)
    * (c) `net.group.dtypes`: `name: str`, `element_type: str`, `element_index: int`, `element_column: OPTIONAL[str]` (as @rbolgaryn  proposed)
* (3) **group assignment in net.gen, net.sgen, ... dataframes**
    * (a) `group: list[str]` in net[elm] dataframe
    * (b) one column per group with dtype `bool`. (which column name? f'group_{group_name}'?

**Comment by rbolgaryn** [comment]
grafik

**Comment by SteffenMeinecke** [comment]
Decision: At least as long as the "element_column functionality" is needed, 2b is the best compromise between functional performance, data amount, (visual) clarity. In accordance with https://github.com/e2nIEE/pandapower/issues/1630, the columns in `net.group` should be named `element_type`, `element`, `reference_column`.

**Comment by SteffenMeinecke** [comment]
@rbolgaryn do you think the failing tests are related to your postresql changes? Seems a bit weird why these test should be affected by the totally new feature group.

**Review by rbolgaryn — APPROVED** [review]

**Pull request merged**